crates/sparrow-runtime/src/read/parquet_file.rs

```diff
@@ -145,55 +176,37 @@ impl parquet::arrow::async_reader::AsyncFileReader for ParquetFile {
         '_,
         parquet::errors::Result<Arc<parquet::file::metadata::ParquetMetaData>>,
     > {
-        futures::future::ready(Ok(self.metadata.clone())).boxed()
+        futures::future::ready(Ok(self.parquet_metadata.clone())).boxed()
     }
 }
 
-/// Fetch the parquet metadata for the given path.
+/// Number of bytes to prefetch of the footer.
 ///
-/// # Optimization Opportunity
-/// This does two `get_range` calls -- one for the FOOTER_SIZE bytes to determine
-/// the footer size, then one for the appropriate number of bytes. It may be better
-/// to start with a reasonable guess (a few KB) and fetch that, and only do the second
-/// fetch if the FOOTER_SIZE indicates it wasn't enough. This could cut the number of
-/// object store calls to get metadata in half. Of course, if the metadata is stored in
-/// the database, then the calls wouldn't be needed at all.
+/// Providing a larger than default (8) value allows skipping an extra round
+/// of fetches:
+///
+/// 1. Fetch last `PREFETCH_FOOTER_BYTES`
+/// 2. Look at last 8 bytes to determine size of footer.
+/// 3. If the footer is larger than `PREFETCH_FOOTER_BYTES` fetch more.
+const PREFETCH_FOOTER_BYTES: usize = 1024;
+
+/// Fetch the parquet metadata for the given path.
 async fn get_parquet_metadata(
     object_store: &dyn ObjectStore,
-    path: &object_store::path::Path,
+    object_meta: &ObjectMeta,
 ) -> error_stack::Result<Arc<ParquetMetaData>, Error> {
-    // First determine the length of the file.
-    let metadata = object_store
-        .head(path)
-        .await
-        .into_report()
-        .change_context(Error::InvalidParquetMetadata)?;
-    let length = metadata.size;
-
-    let footer_bytes = object_store
-        .get_range(path, length - FOOTER_SIZE..length)
-        .await
-        .into_report()
-        .change_context(Error::InvalidParquetMetadata)?;
-    let footer_bytes: &[u8; FOOTER_SIZE] = footer_bytes
-        .as_ref()
-        .try_into()
-        .expect("Footer bytes should have been FOOTER_SIZE");
-    let metadata_len = parquet::file::footer::decode_footer(footer_bytes)
-        .into_report()
-        .change_context(Error::InvalidParquetMetadata)?;
-
-    let metadata_bytes = object_store
-        .get_range(
-            path,
-            length - metadata_len - FOOTER_SIZE..length - FOOTER_SIZE,
-        )
-        .await
-        .into_report()
-        .change_context(Error::InvalidParquetMetadata)?;
-
-    parquet::file::footer::decode_metadata(metadata_bytes.as_ref())
-        .into_report()
-        .map(Arc::new)
-        .change_context(Error::InvalidParquetMetadata)
+    let fetch = |byte_range| {
+        object_store
+            .get_range(&object_meta.location, byte_range)
+            .map_err(|e| parquet::errors::ParquetError::External(Box::new(e)))
+    };
+    parquet::arrow::async_reader::fetch_parquet_metadata(
```

Comment:
Simplify code for fetching (and implement the "prefetching") by using a new helper method from the parquet crate.